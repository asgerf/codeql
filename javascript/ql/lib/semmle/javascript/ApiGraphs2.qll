import javascript
private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
private import internal.CachedStages
private import semmle.javascript.DeepFlow
private import semmle.javascript.dataflow.internal.DataFlowNode

module API2 {
  deprecated Node root() { none() }

  /**
   * A class for contributing new steps for tracking uses of an API.
   */
  class AdditionalUseStep extends Unit {
    /**
     * Holds if use nodes should flow from `pred` to `succ`.
     */
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) { none() }
  }

  private module AdditionalUseStep {
    pragma[nomagic]
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) {
      any(AdditionalUseStep st).step(pred, succ)
    }
  }

  cached
  Node moduleImport(string name) {
    result = DataFlow::moduleImport(name)
    or
    result = DataFlow::moduleMember(name, "default")
  }

  module Node {
    Node ofType(string packageName, string typeName) {
      result.(DataFlow::SourceNode).hasUnderlyingType(packageName, typeName)
    }
  }

  class InvokeNode extends DataFlow::InvokeNode {
    pragma[inline]
    SinkNode getParameter(int i) { result = this.getArgument(i) }

    pragma[inline]
    SinkNode getAParameter() { result = this.getAnArgument() }

    pragma[inline]
    SinkNode getLastParameter() { result = this.getLastArgument() }

    pragma[inline]
    Node getReturn() { result = this }

    pragma[inline]
    Node getInstance() { result = this and this instanceof DataFlow::NewNode }
  }

  class CallNode extends InvokeNode, DataFlow::CallNode { }

  class NewNode extends InvokeNode, DataFlow::NewNode { }

  class Node extends TNode instanceof DataFlow::SourceNode {
    deprecated Node getASuccessor() { none() }

    string toString() { result = super.toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode forward() {
      Deep::hasFlowTo(pragma[only_bind_out](this), pragma[only_bind_into](result))
    }

    pragma[inline]
    DataFlow::Node getAValueReachableFromSource() { result = this.forward().getALocalUse() }

    pragma[inline]
    DataFlow::SourceNode asSource() { result = this }

    pragma[inline]
    API2::CallNode getACall() { result = this.forward().getACall() }

    pragma[inline]
    API2::CallNode getMaybePromisifiedCall() {
      result = this.getACall()
      or
      result = Deep::getABoundInvocation(this, true, _)
    }

    pragma[inline]
    API2::NewNode getAnInstantiation() { result = this.forward().getAnInstantiation() }

    pragma[inline]
    API2::InvokeNode getAnInvocation() { result = this.forward().getAnInvocation() }

    pragma[inline]
    API2::Node getMember(string name) { result = Deep::getLoad(pragma[only_bind_out](this), name) }

    pragma[inline]
    API2::Node getUnknownMember() {
      DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).forward(),
        pragma[only_bind_into](result)) and
      pragma[only_bind_out](result) instanceof DataFlow::PropRead
      // Note: The type check above is not strictly needed since API::Node is a subtype of SourceNode, and PropWrite is not a SourceNode.
      // TODO: Investigate which check is more efficient in practice.
    }

    pragma[inline]
    API2::Node getAMember() { result = pragma[only_bind_out](this.forward()).getAPropertyRead() }

    pragma[inline]
    API2::Node getInstance() {
      result = this.getAnInstantiation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::SinkNode getParameter(int i) {
      Deep::argumentPassing(pragma[only_bind_out](this), i, result)
    }

    pragma[inline]
    API2::SinkNode getLastParameter() {
      result = pragma[only_bind_out](this.getACall()).getLastArgument()
      or
      result =
        TApiSyntheticCallbackArg(Deep::getABoundInvocation(pragma[only_bind_out](this), true, _))
    }

    pragma[inline]
    API2::SinkNode getAParameter() { Deep::argumentPassing(pragma[only_bind_out](this), _, result) }

    pragma[inline]
    API2::SinkNode getReceiver() {
      Deep::receiverPassing(this, result) // TODO: change to only return explicit receivers?
    }

    pragma[inline]
    API2::Node getReturn() {
      result = this.getAnInvocation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::Node getPromised() { result = Deep::getPromised(this) }

    pragma[inline]
    API2::Node getPromisedError() { result = Deep::getPromisedError(this) }

    Node getADecoratedClass() { none() } // TODO

    Node getADecoratedMember() { none() } // TODO

    SinkNode getADecoratedParameter() { none() } // TODO

    /**
     * DEPRECATED. Use `getACall().getNumArgument()` instead, but note that each call may have a different number of arguments.
     *
     * This predicate returns the largest arity among these calls, or zero if there are no calls.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::Node asSink() { none() }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::SourceNode getAValueReachingSink() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource`. */
    deprecated DataFlow::SourceNode getAnImmediateUse() { result = this.asSource() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource`. */
    deprecated DataFlow::Node getAUse() { result = this.getAValueReachableFromSource() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class BacktrackingNode extends TNode instanceof DataFlow::SourceNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode backward() {
      Deep::hasFlowTo(pragma[only_bind_into](result), pragma[only_bind_out](this))
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() { result = this.backward() }

    pragma[inline]
    DataFlow::FunctionNode getAFunction() { result = this.backward() }

    pragma[inline]
    API2::SinkNode getMember(string name) {
      // TODO: include store edges for class members
      // TODO: include reverse-load edges for stepping through long access paths
      result = this.backward().getAPropertyWrite(name).getRhs()
    }

    pragma[inline]
    API2::SinkNode getUnknownMember() {
      exists(DataFlow::PropWrite write |
        DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).backward(),
          pragma[only_bind_into](write)) and
        result = write.getRhs()
      )
    }

    pragma[inline]
    API2::SinkNode getAMember() {
      // TODO: include store edges for class members
      result = this.backward().getAPropertyWrite().getRhs()
    }

    pragma[inline]
    API2::Node getInstance() { result = Deep::getClassReceiverRef(this.backward()) }

    pragma[inline]
    API2::Node getParameter(int i) { Deep::parameterDef(this.backward(), i, result) }

    pragma[inline]
    API2::Node getAParameter() { Deep::parameterDef(this.backward(), _, result) }

    pragma[inline]
    API2::Node getLastParameter() {
      // TODO: is this the best way to do this?
      exists(DataFlow::SourceNode src | src = this.backward() |
        result = src.(DataFlow::FunctionNode).getLastParameter()
        or
        result = src.(DataFlow::ClassNode).getConstructor().getLastParameter()
      )
    }

    pragma[inline]
    API2::Node getReceiver() { result = this.getAFunction().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = Deep::getPrettyReturn(this.getAFunction()) }

    pragma[inline]
    API2::SinkNode getPromised() {
      result = Deep::getStoreRhs(this.backward(), Promises::valueProp())
    }

    pragma[inline]
    API2::SinkNode getPromisedError() {
      result = Deep::getStoreRhs(this.backward(), Promises::errorProp())
    }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /**
     * DEPRECATED. This predicate should only be used on `API::SinkNode`. Once backtracking is performed the node is rarely
     * useful as a sink.
     *
     * Returns the receiver itself as a data-flow node, though this does not generally have the appropriate location for u
     */
    deprecated DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class SinkNode extends TDataFlowNodeOrApiNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    API2::BacktrackingNode backtrack() {
      result = pragma[only_bind_out](this).(DataFlow::Node).getALocalSource()
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() {
      result = this.backtrack().getAValueReachingSink()
    }

    pragma[inline]
    API2::SinkNode getMember(string name) { result = this.backtrack().getMember(name) }

    pragma[inline]
    API2::SinkNode getUnknownMember() { result = this.backtrack().getUnknownMember() }

    pragma[inline]
    API2::SinkNode getAMember() { result = this.backtrack().getAMember() }

    pragma[inline]
    API2::Node getInstance() { result = this.backtrack().getInstance() }

    pragma[inline]
    API2::Node getParameter(int i) {
      result = this.backtrack().getParameter(i)
      or
      Deep::promisifiedCallbackParameterDef(this, i, result)
    }

    pragma[inline]
    API2::Node getAParameter() { result = this.backtrack().getAParameter() }

    pragma[inline]
    API2::Node getLastParameter() { result = this.backtrack().getLastParameter() }

    pragma[inline]
    API2::Node getReceiver() { result = this.backtrack().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = this.backtrack().getReturn() }

    pragma[inline]
    API2::SinkNode getPromised() { result = this.backtrack().getPromised() }

    pragma[inline]
    API2::SinkNode getPromisedError() { result = this.backtrack().getPromisedError() }

    /**
     * Returns this node as a data-flow node.
     */
    pragma[inline]
    DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }
}
