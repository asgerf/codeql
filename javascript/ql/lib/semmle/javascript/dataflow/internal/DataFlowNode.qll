/**
 * INTERNAL: Do not use outside the data flow library.
 *
 * Contains the raw data type underlying `DataFlow::Node`.
 */

private import javascript
private import semmle.javascript.dataflow2.DataFlowImplSpecific::Private as DataFlowPrivate
private import semmle.javascript.dataflow2.FlowSummaryImpl as FlowSummaryImpl
private import semmle.javascript.dataflow2.FlowSummaryImplSpecific as FlowSummaryImplSpecific

/**
 * The raw data type underlying `DataFlow::Node`.
 */
cached
newtype TNode =
  TValueNode(AST::ValueNode nd) or
  TSsaDefNode(SsaDefinition d) or
  TCapturedVariableNode(LocalVariable v) { v.isCaptured() } or
  TPropNode(@property p) or
  TRestPatternNode(DestructuringPattern dp, Expr rest) { rest = dp.getRest() } or
  TElementPatternNode(ArrayPattern ap, Expr p) { p = ap.getElement(_) } or
  TElementNode(ArrayExpr arr, Expr e) { e = arr.getAnElement() } or
  TReflectiveCallNode(MethodCallExpr ce, string kind) {
    ce.getMethodName() = kind and
    (kind = "call" or kind = "apply")
  } or
  TThisNode(StmtContainer f) { f.(Function).getThisBinder() = f or f instanceof TopLevel } or
  TFunctionSelfReferenceNode(Function f) or
  TDestructuredModuleImportNode(ImportDeclaration decl) {
    exists(decl.getASpecifier().getImportedName())
  } or
  THtmlAttributeNode(HTML::Attribute attr) or
  TFunctionReturnNode(Function f) or
  TAsyncFunctionIntermediateStoreReturnNode(Function f) { f.isAsync() } or
  TExceptionalFunctionReturnNode(Function f) or
  TExceptionalInvocationReturnNode(InvokeExpr e) or
  TGlobalAccessPathRoot() or
  TTemplatePlaceholderTag(Templating::TemplatePlaceholderTag tag) or
  TReflectiveParametersNode(Function f) or
  TExprPostUpdateNode(Expr e) {
    e = any(InvokeExpr invoke).getAnArgument() or
    e = any(PropAccess access).getBase() or
    e = any(InvokeExpr invoke).getCallee()
  } or
  TConstructorThisArgumentNode(InvokeExpr e) { e instanceof NewExpr or e instanceof SuperCall } or
  TFlowSummaryNode(FlowSummaryImpl::Private::SummaryNode sn) or
  TFlowSummaryIntermediateAwaitStoreNode(FlowSummaryImpl::Private::SummaryNode sn) {
    FlowSummaryImpl::Private::Steps::summaryStoreStep(sn, DataFlowPrivate::MkAwaited(), _)
  }

/**
 * A data-flow node that is not a flow summary node.
 *
 * This node exists to avoid an unwanted dependency on flow summaries in some parts of the codebase
 * that should not depend on them.
 *
 * In particular, this dependency chain must not result in negative recursion:
 * - Flow summaries can only be created after pruning irrelevant flow summaries
 * - To prune irrelevant flow summaries, we must know which packages are imported
 * - To know which packages are imported, module systems must be evaluated
 * - The AMD and NodeJS module systems rely on data flow to find calls to `require` and similar.
 *   These module systems must therefore use `TEarlyStageNode` instead of `DataFlow::Node`.
 */
class TEarlyStageNode =
  TValueNode or TSsaDefNode or TCapturedVariableNode or TPropNode or TRestPatternNode or
      TElementPatternNode or TElementNode or TReflectiveCallNode or TThisNode or
      TFunctionSelfReferenceNode or TDestructuredModuleImportNode or THtmlAttributeNode or
      TFunctionReturnNode or TExceptionalFunctionReturnNode or TExceptionalInvocationReturnNode or
      TGlobalAccessPathRoot or TTemplatePlaceholderTag or TReflectiveParametersNode or
      TExprPostUpdateNode or TConstructorThisArgumentNode;
