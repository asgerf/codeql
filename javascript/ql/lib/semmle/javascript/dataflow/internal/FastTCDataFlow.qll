private import javascript
private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
private import semmle.javascript.dataflow.internal.StepSummary
private import semmle.javascript.dataflow.TypeTracking

/**
 * Gets a node that is reachable by type-tracking with a TypeTracker that has the given `prop` and `call` state.
 */
pragma[nomagic]
private DataFlow::SourceNode getANodeWithProperty(string prop, boolean call) {
  call = false and
  StepSummary::step(_, result, StoreStep(prop))
  or
  exists(StepSummary step, string prevProp, boolean prevCall |
    StepSummary::step(getANodeWithProperty(prevProp, prevCall), result, step)
  |
    step = LevelStep() and
    call = prevCall and
    prop = prevProp
    or
    step = CallStep() and
    call = true and
    prop = prevProp
    or
    step = ReturnStep() and
    prevCall = false and
    call = false and
    prop = prevProp
    or
    step = CopyStep(prevProp) and
    prop = prevProp and
    call = prevCall
    or
    step = LoadStoreStep(prevProp, prop) and
    call = prevCall
    or
    exists(PropertySet props |
      step = WithoutPropStep(props) and
      not props.getAProperty() = prevProp and
      call = prevCall and
      prop = prevProp
    )
  )
}

private newtype TIntermediateNode =
  MkIntermediateNode(DataFlow::SourceNode node, string prop) {
    node = getANodeWithProperty(prop, _)
    or
    prop = "" and exists(node)
  }

private class IntermediateNode extends TIntermediateNode {
  private string getProperty() { this = MkIntermediateNode(_, result) }

  private DataFlow::SourceNode getNode() { this = MkIntermediateNode(result, _) }

  string toString() { result = this.getNode() + " [" + this.getProperty() + "]" }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this.getNode().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }
}

pragma[noopt]
private predicate commonEdge(IntermediateNode pred, IntermediateNode succ) {
  exists(
    DataFlow::SourceNode predNode, DataFlow::SourceNode succNode, StepSummary step, string prop
  |
    StepSummary::step(predNode, succNode, step)
  |
    step = LoadStep(prop) and
    pred = MkIntermediateNode(predNode, prop) and
    succ = MkIntermediateNode(succNode, "")
    or
    step = StoreStep(prop) and
    pred = MkIntermediateNode(predNode, "") and
    succ = MkIntermediateNode(succNode, prop)
    or
    step = CopyStep(prop) and
    pred = MkIntermediateNode(predNode, prop) and
    succ = MkIntermediateNode(succNode, prop)
    or
    exists(string prop2 |
      step = LoadStoreStep(prop, prop2) and
      pred = MkIntermediateNode(predNode, prop) and
      succ = MkIntermediateNode(succNode, prop2)
    )
    or
    exists(PropertySet props |
      step = WithoutPropStep(props) and
      pred = MkIntermediateNode(predNode, prop) and
      succ = MkIntermediateNode(succNode, prop) and
      not prop = props.getAProperty()
    )
    or
    step = LevelStep() and
    predNode != succNode and
    pred = MkIntermediateNode(predNode, prop) and
    succ = MkIntermediateNode(succNode, prop)
  )
}

cached
private module Cached {
  pragma[noopt]
  cached
  predicate stage1Edge(IntermediateNode pred, IntermediateNode succ) {
    commonEdge(pred, succ)
    or
    exists(
      DataFlow::SourceNode predNode, DataFlow::SourceNode succNode, StepSummary step, string prop
    |
      StepSummary::step(predNode, succNode, step) and
      step = ReturnStep() and
      pred = MkIntermediateNode(predNode, prop) and
      succ = MkIntermediateNode(succNode, prop)
    )
  }

  pragma[noopt]
  cached
  predicate stage2Edge(IntermediateNode pred, IntermediateNode succ) {
    commonEdge(pred, succ)
    or
    exists(
      DataFlow::SourceNode predNode, DataFlow::SourceNode succNode, StepSummary step, string prop
    |
      StepSummary::step(predNode, succNode, step) and
      step = CallStep() and
      pred = MkIntermediateNode(predNode, prop) and
      succ = MkIntermediateNode(succNode, prop)
    )
  }
}

private import Cached

pragma[nomagic]
predicate rawNode(DataFlow::SourceNode node, IntermediateNode inode) {
  inode = MkIntermediateNode(node, "")
}

bindingset[node]
pragma[inline_late]
pragma[noopt]
DataFlow::SourceNode getAGlobalSuccessor(DataFlow::SourceNode node) {
  exists(IntermediateNode start, IntermediateNode mid, IntermediateNode end |
    rawNode(node, start) and
    stage1Edge*(start, mid) and
    stage2Edge*(mid, end) and
    rawNode(result, end)
  )
}

bindingset[node]
pragma[inline_late]
pragma[noopt]
DataFlow::SourceNode getAGlobalPredecessor(DataFlow::SourceNode node) {
  exists(IntermediateNode start, IntermediateNode mid, IntermediateNode end |
    rawNode(node, end) and
    stage2Edge*(mid, end) and
    stage1Edge*(start, mid) and
    rawNode(result, start)
  )
}

/**
 * INTERNAL. DO NOT USE.
 *
 * Used for testing.
 */
private module Internal {
  pragma[nomagic]
  private DataFlow::SourceNode getAGlobalSuccessorMaterialized(DataFlow::SourceNode node) {
    result = getAGlobalSuccessor(node)
  }

  private DataFlow::SourceNode trackNode(DataFlow::SourceNode src, DataFlow::TypeTracker t) {
    t.start() and
    result = src
    or
    exists(DataFlow::TypeTracker t2 | result = trackNode(src, t2).track(t2, t))
  }

  pragma[nomagic]
  private DataFlow::SourceNode trackNode(DataFlow::SourceNode src) {
    result = trackNode(src, DataFlow::TypeTracker::end())
  }

  pragma[noopt]
  query predicate diffWithTypeTracking(
    DataFlow::SourceNode src, DataFlow::SourceNode dst, string diff
  ) {
    dst = getAGlobalSuccessorMaterialized(src) and
    not dst = trackNode(src) and
    diff = "gained"
    or
    dst = trackNode(src) and
    not dst = getAGlobalSuccessorMaterialized(src) and
    diff = "lost"
  }
}
