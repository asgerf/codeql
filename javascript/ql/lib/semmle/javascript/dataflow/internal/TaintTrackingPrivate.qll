private import javascript
private import semmle.javascript.dataflow.internal.DataFlowPrivate
private import semmle.javascript.dataflow.internal.Contents::Public
private import semmle.javascript.dataflow.internal.sharedlib.FlowSummaryImpl as FlowSummaryImpl

cached
predicate defaultAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
  TaintTracking::AdditionalTaintStep::step(node1, node2)
  or
  FlowSummaryImpl::Private::Steps::summaryLocalStep(node1.(FlowSummaryNode).getSummaryNode(),
    node2.(FlowSummaryNode).getSummaryNode(), false)
  or
  // Convert steps into and out of array elements to plain taint steps
  FlowSummaryImpl::Private::Steps::summaryReadStep(node1.(FlowSummaryNode).getSummaryNode(),
    ContentSet::arrayElement(), node2.(FlowSummaryNode).getSummaryNode())
  or
  FlowSummaryImpl::Private::Steps::summaryStoreStep(node1.(FlowSummaryNode).getSummaryNode(),
    ContentSet::arrayElement(), node2.(FlowSummaryNode).getSummaryNode())
}

/**
 * Holds if `node` should be a sanitizer in all global taint flow configurations
 * but not in local taint.
 */
cached
predicate defaultTaintSanitizer(DataFlow::Node node) {
  node instanceof DataFlow::VarAccessBarrier or
}
/**
 * Holds if default taint-tracking should allow implicit reads
 * of `c` at sinks and inputs to additional taint steps.
 */
bindingset[node]
predicate defaultImplicitTaintRead(DataFlow::Node node, ContentSet c) {
  exists(node) and
  c = ContentSet::promiseValue()
}
