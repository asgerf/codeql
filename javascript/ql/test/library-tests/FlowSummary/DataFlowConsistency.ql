import javascript
import semmle.javascript.dataflow2.DataFlowImplSpecific::Private
import semmle.javascript.dataflow2.DataFlowImplConsistency::Consistency
import semmle.javascript.dataflow.internal.DataFlowNode

private predicate isAmbientNode(DataFlow::Node node) {
  exists(AstNode n | n.isAmbient() |
    node = TValueNode(n) or
    node = TThisNode(n) or
    node = TReflectiveParametersNode(n) or
    node = TPropNode(n) or
    node = TFunctionSelfReferenceNode(n) or
    node = TAsyncFunctionIntermediateStoreReturnNode(n) or
    node = TExceptionalFunctionReturnNode(n) or
    node = TExprPostUpdateNode(n) or
    node = TExceptionalInvocationReturnNode(n) or
    node = TDestructuredModuleImportNode(n)
  )
}

class Config extends ConsistencyConfiguration {
  override predicate missingLocationExclude(DataFlow::Node n) {
    n instanceof FlowSummaryNode
    or
    n instanceof FlowSummaryIntermediateAwaitStoreNode
    or
    n = DataFlow::globalAccessPathRootPseudoNode()
  }

  override predicate uniqueNodeLocationExclude(DataFlow::Node n) { this.missingLocationExclude(n) }

  override predicate uniqueEnclosingCallableExclude(DataFlow::Node n) { isAmbientNode(n) }

  override predicate uniqueCallEnclosingCallableExclude(DataFlowCall call) {
    isAmbientNode(call.asOrdinaryCall()) or
    isAmbientNode(call.asAccessorCall())
  }
}
