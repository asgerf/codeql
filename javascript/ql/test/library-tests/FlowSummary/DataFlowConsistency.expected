uniqueEnclosingCallable
uniqueCallEnclosingCallable
uniqueType
uniqueNodeLocation
missingLocation
uniqueNodeToString
| file://:0:0:0:0 | (no string representation) | Node should have one toString but has 0. |
| file://:0:0:0:0 | (no string representation) | Node should have one toString but has 0. |
| file://:0:0:0:0 | (no string representation) | Node should have one toString but has 0. |
| file://:0:0:0:0 | (no string representation) | Node should have one toString but has 0. |
missingToString
| Nodes without toString: 4 |
parameterCallable
localFlowIsLocal
readStepIsLocal
storeStepIsLocal
compatibleTypesReflexive
unreachableNodeCCtx
localCallNodes
postIsNotPre
postHasUniquePre
uniquePostUpdate
postIsInSameCallable
reverseRead
| tst.js:84:14:84:40 | t.flowI ... urce()) | Origin of readStep is missing a PostUpdateNode. |
| tst.js:86:14:86:59 | t.flowI ... rce())) | Origin of readStep is missing a PostUpdateNode. |
argHasPostUpdate
| tst.js:16:32:16:39 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:17:33:17:40 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:21:14:21:21 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:41:14:41:31 | { prop: source() } | ArgumentNode is missing PostUpdateNode. |
| tst.js:41:22:41:29 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:128:18:128:25 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:133:18:133:25 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:142:5:142:31 | t.flowI ... urce()) | ArgumentNode is missing PostUpdateNode. |
| tst.js:143:5:143:12 | source() | ArgumentNode is missing PostUpdateNode. |
| tst.js:144:5:144:10 | "safe" | ArgumentNode is missing PostUpdateNode. |
| tst.js:145:12:145:23 | [x1, x2, x3] | ArgumentNode is missing PostUpdateNode. |
| tst.js:154:22:154:29 | source() | ArgumentNode is missing PostUpdateNode. |
postWithInFlow
| tst.js:93:24:93:74 | new Pro ... rce())) | PostUpdateNode should not be the target of local flow. |
| tst.js:96:3:96:53 | new Pro ... rce())) | PostUpdateNode should not be the target of local flow. |
| tst.js:97:3:97:53 | new Pro ... rce())) | PostUpdateNode should not be the target of local flow. |
| tst.js:98:3:98:52 | new Pro ... rce())) | PostUpdateNode should not be the target of local flow. |
| tst.js:99:3:99:52 | new Pro ... rce())) | PostUpdateNode should not be the target of local flow. |
viableImplInCallContextTooLarge
uniqueParameterNodeAtPosition
uniqueParameterNodePosition
uniqueContentApprox
